### DISCUSSION TITLE
How to plot a curved arrow?

### DISCUSSION THREAD

**Discussion by huisky2000** [pyvista/pyvista#5023] [Q&A] [answered]
Hi,
I see  that we can plot an 'normal' arrow by using 'pv.Arrow'. I didn't find the option for plotting a curved arrow. I'd like to show the applied force and moments to an object. The moments should be presented with curved arrow.
Thanks!
Regards
Huisky

**Comment by banesullivan** [comment]
Sounds like you want something similar to creating streamlines through a vector field. Maybe this helps: 

- https://docs.pyvista.org/version/stable/examples/01-filter/streamlines.html#streamlines
- https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.streamlines.html
- https://docs.pyvista.org/version/stable/examples/01-filter/streamlines_2D.html

**Comment by huisky2000** [comment]
Hi,
Thanks a lot for the follow-up. It is not the streamlines that I want to create. I want to make a similar plot as 'pv.Arrow', but it is with a curved arrow, something like this
https://i.redd.it/zopr70107gjy.png
So I can make some illustrations for a 3d solid under force and moment loads using Pyvista. Is it possible to implement this?
Regards
Huisky

**Reply by banesullivan** [reply]
Oohh, thanks for providing an example -- very different from what I was imagining. You need to create the loop as a polyline and then the arrowhead as a cone separately. I wish I had the time to try to put this together for you but I hope that helps.

I recommend making a function that will create a non-closed loop of points given a center point and radius. Then create a `pv.Cone()` to center on the last point of the line. Then you can merge these as one "mesh" and translate/rotate as needed. 

To create the polyline, ref: https://docs.pyvista.org/version/stable/examples/00-load/create-spline.html

**Reply by darikg** [reply, marked as answer]
Fun little challenge. 

```python
import pyvista as pv
import numpy as np


# Feel like this might already be available in pyvista somewhere, but I can't find it
def rotate_to(a, b):
    """Returns a rotation matrix taking the vector A to B"""
    # Naive approach is unstable if a and b are near parallel
    theta = np.arccos(np.dot(a, b))
    eps = 1e-3
    if np.absolute(theta) < eps:
        return np.eye(4)
    elif np.absolute(np.pi - theta) < eps:  # Close to 180 degrees
        # Choose the coordinate axis most orthogonal to A
        x = np.zeros(3)
        x[np.argmin(np.absolute(a))] = 1.0
        axis = np.cross(a, x)
        axis /= np.linalg.norm(axis)
        return pv.utilities.transformations.axis_angle_rotation(axis, theta, deg=False)
    else:
        axis = np.cross(a, b)
        return pv.utilities.transformations.axis_angle_rotation(axis, theta, deg=False)


def semi_circular_arrow(
        start_angle=0,
        circ_frac=0.75,
        body_axial_res=100,
        body_radial_res=50,
        head_radial_res=100,
        circ_radius=10,
        body_radius=1,
        head_length=3,
        head_radius_frac=1.5,
        normal=None,
        center=None,
):

    t = np.linspace(0, circ_frac * 2 * np.pi, body_axial_res) + start_angle
    x = circ_radius * np.cos(t)
    y = circ_radius * np.sin(t)
    z = np.zeros(body_axial_res)
    body_pts = np.column_stack([x, y, z])
    body = pv.MultipleLines(body_pts).tube(body_radius, n_sides=body_radial_res)

    # Direction the head points
    dhead = body_pts[-1] - body_pts[-2]
    dhead /= np.linalg.norm(dhead)

    head = pv.Cone(
        center=body_pts[-1] + dhead * (head_length / 2),
        direction=dhead,
        height=head_length,
        radius=head_radius_frac * body_radius,
        resolution=head_radial_res,
    )

    arrow = body.merge(head, merge_points=False)

    if normal is not None:
        arrow = arrow.transform(rotate_to((0, 0, 1), normal), inplace=False)

    if center is not None:
        arrow = arrow.translate(center, inplace=False)

    return arrow


p = pv.Plotter()
center = (1, 1, 1)
normal = (1, 0, 0)
p.add_mesh(semi_circular_arrow(center=center, normal=normal), color='blue')
p.add_mesh(pv.Arrow(start=center, direction=normal, scale=10), color='red')
p.show()
```

![image](https://github.com/pyvista/pyvista/assets/6875882/1436afc1-c851-47a9-9684-bfa7f8be729b)

**Reply by banesullivan** [reply]
Nice work, @darikg!!

**Reply by huisky2000** [reply]
Thanks a lot and nice work!
